### PR TITLE
Implement `Debug` for public types

### LINKED ISSUE #427: Public types should implement `Debug`
This can be linted with `#![deny(missing_debug_implementations)]` and then basically hitting "go" until everything compiles again!

### PR DESCRIPTION
**Pull request by tmccombs**
Fixes #427.

I included what I though was important in the implementations, but I'm open to suggestions.

### CODE REVIEW

**Comment by nikomatsakis** [comment]
@tmccombs hmm -- I'm seeing a bunch of travis failures! do you mind taking a look?

**Comment by KodrAus** [comment]
That'll be the `deny(missing_debug)`, which includes `pub` structs even if they aren't re-exported.

@nikomatsakis would you rather keep the lint and work around `pub(not pub)` structs or just remove it?

**Comment by tmccombs** [comment]
Ah I didn't run with rayon_unstable feature turned on locally. I'll look into them.

**Comment by tmccombs** [comment]
hmm, I had a typo, so it didn't actually get everything, I'll work on getting all the structs I missed.

File: rayon-futures/src/lib.rs [hunk 1]
```diff
@@ -113,6 +115,11 @@ impl<T, E> Drop for RayonFuture<T, E> {
     }
 }
 
+impl<T, E> fmt::Debug for RayonFuture<T, E> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.write_str("RayonFuture(...)")
```
Comment (KodrAus) [inline rayon-futures/src/lib.rs]: This could probably just be `fmt.debug_struct("RayonFuture").finish()`.

File: rayon-core/src/scope/mod.rs [hunk 2]
```diff
@@ -346,3 +347,14 @@ impl<'scope> Scope<'scope> {
...
     }
 }
+
+impl<'scope> fmt::Debug for Scope<'scope> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        use latch::LatchProbe;
+        fmt.debug_struct("Scope")
+            .field("owner_thread", &self.owner_thread)
+            .field("panic", &self.panic)
+            .field("job_completed", &self.job_completed_latch.probe())
```
Comment (cuviper) [inline rayon-core/src/scope/mod.rs]: I think `probe()` will always be true until the scope is out of the user's hands.  For debugging purposes, the actual counter value would be more interesting.  How about adding `Debug` to `CountLatch` and show that here?

Comment (tmccombs) [reply rayon-core/src/scope/mod.rs]: alright

**Comment by cuviper** [comment]
I added a couple more that hit master in the meantime.

Thanks!

**Pull request merged**